File: src/ros2_bridge/mappers/geometry_msgs/transform.rs
//! Typed mapper for `geometry_msgs/msg/Transform`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn transform_to_bus(msg: ros_env::geometry_msgs::msg::Transform) -> crate::geometry_msgs::msg::v1::Transform {
    crate::geometry_msgs::msg::v1::Transform {
        translation: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.translation)),
        rotation: Some(crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_bus(msg.rotation)),
    }
}

pub(crate) fn transform_to_ros(bus: crate::geometry_msgs::msg::v1::Transform) -> ros_env::geometry_msgs::msg::Transform {
    ros_env::geometry_msgs::msg::Transform {
        translation: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(bus.translation.unwrap_or_default()),
        rotation: crate::ros2_bridge::mappers::geometry_msgs::quaternion::quaternion_to_ros(bus.rotation.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsTransformMapper;

impl TypedTopicMapper for GeometryMsgsTransformMapper {
    type Ros = ros_env::geometry_msgs::msg::Transform;
    type Bus = crate::geometry_msgs::msg::v1::Transform;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(transform_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(transform_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/transform.py
"""Generated mapper for `geometry_msgs/msg/Transform`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.vector3 import vector3_to_bus, vector3_to_ros
from robot_bus.ros2_bridge.mappers.geometry_msgs.quaternion import quaternion_to_bus, quaternion_to_ros

def transform_to_bus(msg):
    from robot_bus.geometry_msgs.msg.v1 import Transform as BusMsg

    bus = BusMsg()
    bus.translation.CopyFrom(vector3_to_bus(msg.translation))
    bus.rotation.CopyFrom(quaternion_to_bus(msg.rotation))
    return bus


def transform_to_ros(bus):
    from geometry_msgs.msg import Transform as RosMsg

    out = RosMsg()
    out.translation = vector3_to_ros(bus.translation)
    out.rotation = quaternion_to_ros(bus.rotation)
    return out


class GeometryMsgsTransformMapper:
    def ros_msg_type(self):
        from geometry_msgs.msg import Transform as RosMsg

        return RosMsg

    def ros_to_bus(self, msg) -> bytes:
        return transform_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        from robot_bus.geometry_msgs.msg.v1 import Transform as BusMsg

        bus = BusMsg()
        bus.ParseFromString(payload)
        return transform_to_ros(bus)
